### PR TITLE
Feature/issue 154 routing

### PR DESCRIPTION
**Pull request by bensinjin**
This PR replaces our current router with the very established [react router](https://reacttraining.com/react-router/native/guides/philosophy) these are the same devs that produce the history library which many router solutions leverage. Outlined in #102, #130, and #154 are the issues we were encountering with our existing router. Another major issue is that Redux-router handles history internally which prevented us from dealing with it directly which from what I could tell would be the only way to fix our routing issues.  I decided it would be simpler, quicker, and cleaner to swap out this piece for something I feel will serve us better moving forward.  Please try it out, you should notice:

- Back button works as expected on all pages
- Android back button works as expected and does not close the app
- The system for adding new routes, with or without parameter support, and navigating to them is quick and straight forward to implement, react-router takes care of knowing what to render based on props to the `Route` component
- We have to write far less often redundant routing related code. Any component that requires access to routing now references one interface: `RouterProps` so we don't have to be constantly writing variations of props like `goToTaskDetail` and `goToMyPlan` for component A and then again for component B etc..

### CODE REVIEW

File: src/application/store.ts [hunk 1]
```diff
@@ -24,9 +23,9 @@ export type Store = { readonly applicationState: StoreForApplicationState };
 
 type InitialState = { readonly applicationState: { readonly localeInStore: locale.Store } };
 
-export function buildStore(router: ApplicationRouter, saga: ApplicationSaga): ReturnType<typeof createStore> {
-    const middleware = applyMiddleware(router.middleware, saga.middleware);
-    const reducer = combineReducers({ location: router.reducer, applicationState: reducerForApplicationState });
+export function buildStore(saga: ApplicationSaga): ReturnType<typeof createStore> {
+    const middleware = applyMiddleware(saga.middleware);
+    const reducer = combineReducers({ applicationState: reducerForApplicationState });
```
[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply src/application/store.ts]: This is maybe a bit of a big refactor, so we can do it later if you prefer.

File: package.json [hunk 2]
```diff
@@ -78,6 +76,7 @@
   "dependencies": {
     "@lingui/react": "2.1.0",
     "@types/query-string": "^6.1.0",
+    "@types/react-router-native": "^4.2.3",
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply package.json]: Fixed

File: src/application/__tests__/routing.test.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    it('For Routes.ArticleDetail returns path: "/article/:articleId', () => {
+        expect(routePath(Routes.ArticleDetail)).toEqual('/article/:articleId');
+    });
+});
+
+describe('the routePathWithParameter function', () => {
+
+    it('Replaces a Route argument placeholder with provided value', () => {
+        expect(routePathWithParameter(Routes.LearnDetail, 'l1')).toEqual('/learn/l1');
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/application/__tests__/routing.test.ts]: Yeah I like that, "rejigged"

File: src/application/routing.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        case Routes.Learn:
+            return '/learn';
+        case Routes.LearnDetail:
+            return '/learn/:learnId';
+        case Routes.TaskDetail:
+            return '/task/:taskId';
+        case Routes.ArticleDetail:
+            return '/article/:articleId';
+    }
+};
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/application/routing.ts]: Implemented. Unfortunately I did still have to export routePath just for the 
route definitions in `main_page_switcher.tsx` (all other places use appropriate `routePathWithoutParameter` or `routePathWithParameter`) which I'm not too stoked on. Got any ideas for this without having to duplicate the path definitions?

File: src/application/store.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/application/store.ts]: Of course, fixed

File: src/components/explore/explore_section_connected_component.ts [hunk 5]
```diff
@@ -2,12 +2,13 @@ import { Dispatch } from 'redux';
...
+import { selectLearnByPathParameter } from '../../selectors/explore';
+import { selectTasksForLearnDetail } from '../../selectors/tasks';
+import { RouterProps } from '../../application/routing';
 
-const mapStateToProps = (store: Store): ExploreSectionProps => ({
-    section: selectCurrentExploreSection(store),
-    tasks: selectTasksForCurrentExploreSection(store),
+const mapStateToProps = (store: Store, ownProps: RouterProps): ExploreSectionProps => ({
+    section: selectLearnByPathParameter(store, ownProps.match.params.learnId),
+    tasks: selectTasksForLearnDetail(store, ownProps.match.params.learnId),
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/components/explore/explore_section_connected_component.ts]: Fixed throughout

File: src/application/routing.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/application/routing.ts]: See below - we now use the goTo functions for navigation needs which utilize the path functions.

File: src/components/main/footer.tsx [hunk 6]
```diff
@@ -1,44 +1,41 @@
...
-    readonly goToExplore: () => void;
-}
-
-export const Component: React.StatelessComponent<Props & Actions> = (props: Props & Actions): JSX.Element => {
-    const { goToHome, goToQuestionnaire, goToPlan, goToExplore, routeInProps }: Props & Actions = props;
-
-    if (routeInProps.pageType === store.Page.Welcome) {
+export const FooterComponent: React.StatelessComponent<FooterProps> = (props: FooterProps): JSX.Element => {
+    const path = props.location.pathname;
+    if (path === '/') {
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/components/main/footer.tsx]: Fixed. I think there is a little confusion around `routePath()` so I've attempted to clear it up by changing the name to `routePathDefinition()` as it actually returns the shape of the path not the path value. For example: "task/:taskId"  would equal shape and "task/t1" would equal value. To actually get the path value you would now call either `routePathWithParameter()` or `routePathWithoutParameter()` and that is what we're now using here.

File: src/components/tasks/task_detail_connected_component.tsx [hunk 7]
```diff
@@ -5,19 +5,17 @@ import { TaskDetailProps, TaskDetailActions, TaskDetailComponent, TaskServiceUpd
...
 
 interface StateProps extends TaskDetailProps {
     readonly searchQuery: string;
     readonly taskServices: TaskServices;
 }
 
-function mapStateToProps(store: Store): StateProps {
-    const task: Task  = selectCurrentTask(store);
+function mapStateToProps(store: Store, ownProps: RouterProps): StateProps {
+    const task: Task = selectTaskByPathParameter(store, ownProps.match.params.taskId);
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/components/tasks/task_detail_connected_component.tsx]: Fixed

File: src/components/main/header.tsx [hunk 8]
```diff
@@ -3,36 +3,34 @@ import { Header, Left, Button, Icon, Title, Body, Right } from 'native-base';
...
+    if (props.location.pathname === '/') {
         // tslint:disable-next-line:no-null-keyword
         return null;
     }
 
     return (
         <Header>
             <Left>
-                <Button transparent disabled={!canGoBack} onPress={goBack}>
+                <Button transparent onPress={(): void => props.history.goBack()}>
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/components/main/header.tsx]: It's conceivable conditionally rendering the back back as we currently are will be fine.

**Comment by bensinjin** [comment]
@rasmus-storjohann-PG everything here has been addressed.

File: src/components/articles/article_list_item.tsx [hunk 9]
```diff
@@ -2,18 +2,16 @@ import React from 'react';
...
-    readonly goToArticleDetail: (articleId: Id) => SetArticleDetailPageAction;
-}
-type AllArticleListItemProps = ArticleListItemActions & ArticleListItem;
+type AllArticleListItemProps = ArticleListItem & RouterProps;
 
 export const ArticleListItemComponent: React.StatelessComponent<AllArticleListItemProps> =
-    (props: AllArticleListItemProps): JSX.Element => (
-        <ListItem noIndent noBorder button onPress={(): SetArticleDetailPageAction => props.goToArticleDetail(props.id)}>
+    (props: AllArticleListItemProps): JSX.Element => {
+    const goToArticleDetail = (): void => goToRouteWithParameter(Routes.ArticleDetail, props.id, props.history);
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/components/articles/article_list_item.tsx]: Good point, fixed.

File: src/selectors/explore.ts [hunk 10]
```diff
@@ -1,9 +1,8 @@
-import * as app from '../application/store';
+import { Store } from '../stores';
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/selectors/explore.ts]: I like AppStore.

File: src/application/routing.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
+import { RouteComponentProps } from 'react-router-native';
+import { History } from 'history';
+
+// tslint:disable-next-line:no-any
+export type RouterProps = RouteComponentProps<any>;
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/application/routing.ts]: Fixed, now using proper type.

**Comment by bensinjin** [comment]
@rasmus-storjohann-PG We're good to go.

**Pull request merged**